File: src/main/java/pl/training/module5/extractsubclass/before/DeliveryJob.java
package pl.training.module5.extractsubclass.before;

import java.time.Instant;
import java.util.Objects;
import java.util.Optional;

public final class DeliveryJob {
    private static final String SENT = "SENT";
    private static final String WAITING_UNTIL = "WAITING_UNTIL ";

    private final Optional<Instant> scheduledAt;

    private DeliveryJob(Optional<Instant> scheduledAt) {
        this.scheduledAt = scheduledAt;
    }

    public static DeliveryJob immediate() {
        return new DeliveryJob(Optional.empty());
    }

    public static DeliveryJob scheduled(Instant scheduledAt) {
        return new DeliveryJob(Optional.of(Objects.requireNonNull(
                scheduledAt,
                "scheduledAt must not be null")));
    }

    public String dispatchAt(Instant now) {
        Objects.requireNonNull(now, "now must not be null");

        return scheduledAt
                .filter(now::isBefore)
                .map(instant -> WAITING_UNTIL + instant)
                .orElse(SENT);
    }
}

File: csharp/src/Training.Module5/ExtractSubclass/Before/DeliveryJob.cs
namespace Training.Module5.ExtractSubclass.Before;

public sealed class DeliveryJob
{
    private const string Sent = "SENT";
    private const string WaitingUntil = "WAITING_UNTIL ";

    private readonly DateTimeOffset? _scheduledAt;

    private DeliveryJob(DateTimeOffset? scheduledAt)
    {
        _scheduledAt = scheduledAt;
    }

    public static DeliveryJob Immediate()
    {
        return new DeliveryJob(null);
    }

    public static DeliveryJob Scheduled(DateTimeOffset scheduledAt)
    {
        return new DeliveryJob(scheduledAt);
    }

    public string DispatchAt(DateTimeOffset now)
    {
        return _scheduledAt is { } instant && now < instant
            ? WaitingUntil + InstantFormat.ToIsoString(instant)
            : Sent;
    }
}
